File: blaze_flutter_sdk/android/src/main/kotlin/com/blaze/flutterblazesdk/players/moments/BlazeReactMomentsPlayerStyleMerger.kt
package com.blaze.flutterblazesdk.players.moments

import android.content.Context
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerBodyTextStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerButtonsStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerChipStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerChipsStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerCtaIconStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerCtaStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerFirstTimeSlideInstructionsStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerFirstTimeSlideStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerFooterGradientStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerHeaderGradientStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerHeadingTextStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerSeekBarStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerCustomActionButton
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerFollowEntityStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerFollowEntityStateStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerFollowEntityAvatarStyle
import com.blaze.blazesdk.style.players.moments.BlazeMomentsPlayerFollowEntityChipStyle
import com.blaze.blazesdk.follow.models.BlazeFollowEntityType
import com.blaze.blazesdk.style.shared.models.blazeDp
import com.blaze.flutterblazesdk.players.shared.BlazeReactCustomActionButton
import com.blaze.flutterblazesdk.players.shared.mergeButtonThemes
import com.blaze.flutterblazesdk.players.shared.mergedWith
import com.blaze.flutterblazesdk.players.shared.toNativeParams
import com.blaze.flutterblazesdk.utils.parsing.mergedWith
import com.blaze.flutterblazesdk.utils.parsing.safeParseColor
import com.blaze.flutterblazesdk.utils.parsing.toColorResId
import com.blaze.flutterblazesdk.utils.parsing.toFontResId
import com.blaze.flutterblazesdk.utils.parsing.toImageResId

fun BlazeMomentsPlayerStyle.mergedWith(
    customization: BlazeReactMomentsPlayerStyle?,
    context: Context
): BlazeMomentsPlayerStyle {
    customization ?: return this

    val merged = this

    merged.headingText = this.headingText.mergedWith(customization.headingText, context)
    merged.bodyText = this.bodyText.mergedWith(customization.bodyText, context)
    merged.buttons = this.buttons.mergedWith(customization.buttons, context)
    merged.chips = this.chips.mergedWith(customization.chips)
    merged.cta = this.cta.mergedWith(customization.cta, context)
    safeParseColor(customization.backgroundColor)?.let { merged.backgroundColor = it }
    merged.headerGradient = this.headerGradient.mergedWith(customization.headerGradient)
    merged.footerGradient = this.footerGradient.mergedWith(customization.footerGradient)
    merged.seekBar = this.seekBar.mergedWith(customization.seekBar, context)
    merged.firstTimeSlide = this.firstTimeSlide.mergedWith(customization.firstTimeSlide, context)
    customization.bottomComponentsAlignment?.let {
        merged.bottomComponentsAlignment = it.mapToBlazeEnumClass()
    }
    customization.playerDisplayMode?.let { merged.playerDisplayMode = it.mapToBlazeEnumClass() }
    merged.followEntity = this.followEntity.mergedWith(customization.followEntity)
    return merged
}

fun BlazeMomentsPlayerFollowEntityStyle.mergedWith(
    customization: BlazeReactMomentsPlayerFollowEntityStyle?,
): BlazeMomentsPlayerFollowEntityStyle {
    customization ?: return this

    val merged = this
    merged.isVisible = customization.isVisible ?: this.isVisible
    merged.followState = this.followState.mergedWith(customization.followState)
    merged.unfollowState = this.unfollowState.mergedWith(customization.unfollowState)
    merged.entityType = mergeEntityType(customization.entityType) ?: this.entityType
    return merged
}

fun BlazeMomentsPlayerFollowEntityStateStyle.mergedWith(
    customization: BlazeReactMomentsPlayerFollowEntityStateStyle?,
): BlazeMomentsPlayerFollowEntityStateStyle {
    customization ?: return this

    val merged = this
    merged.avatar = this.avatar.mergedWith(customization.avatar)
    merged.chip = this.chip.mergedWith(customization.chip)
    return merged
}

fun BlazeMomentsPlayerFollowEntityAvatarStyle.mergedWith(
    customization: BlazeReactMomentsPlayerFollowEntityAvatarStyle?,
): BlazeMomentsPlayerFollowEntityAvatarStyle {
    customization ?: return this

    val merged = this
    merged.borderWidth = customization.borderWidth?.toInt()?.blazeDp ?: this.borderWidth
    merged.borderColor = safeParseColor(customization.borderColor) ?: this.borderColor
    return merged
}

fun BlazeMomentsPlayerFollowEntityChipStyle.mergedWith(
    customization: BlazeReactMomentsPlayerFollowEntityChipStyle?,
): BlazeMomentsPlayerFollowEntityChipStyle {
    customization ?: return this

    val merged = this
    merged.backgroundColor = safeParseColor(customization.backgroundColor) ?: this.backgroundColor
    merged.iconColor = safeParseColor(customization.iconColor) ?: this.iconColor
    customization.contentSource?.let { merged.contentSource = it.mapToBlazeEnumClass() }
    return merged
}

private fun mergeEntityType(react: BlazeReactFollowEntityType?): BlazeFollowEntityType? {
    react ?: return null
    val fallback = react.fallbackType?.let { mergeEntityType(it) }

    return when (react.type) {
        "FirstAvailable" -> BlazeFollowEntityType.FirstAvailable
        "Player" -> BlazeFollowEntityType.Player(fallbackType = fallback)
        "Team" -> BlazeFollowEntityType.Team(fallbackType = fallback)
        "Property" -> BlazeFollowEntityType.Property(fallbackType = fallback)
        else -> null
    }
}

fun BlazeMomentsPlayerFirstTimeSlideStyle.mergedWith(
    customization: BlazeReactMomentsPlayerFirstTimeSlideStyle?,
    context: Context
): BlazeMomentsPlayerFirstTimeSlideStyle {
    customization ?: return this

    val merged = this
    merged.show = customization.show ?: this.show
    customization.backgroundColor?.colorFileName?.toColorResId(context)?.let {
        merged.backgroundColorResId = it
    }
    merged.cta = this.cta.mergedWith(customization.cta, context)
    merged.mainTitle = this.mainTitle.mergedWith(customization.mainTitle, context)
    merged.subtitle = this.subtitle.mergedWith(customization.subtitle, context)
    merged.instructions = this.instructions.mergedWith(customization.instructions, context)

    return merged
}

fun BlazeMomentsPlayerHeadingTextStyle.mergedWith(
    customization: BlazeReactMomentsPlayerHeadingTextStyle?,
    context: Context
): BlazeMomentsPlayerHeadingTextStyle {
    customization ?: return this

    val merged = this

    merged.textSize = customization.textSize ?: this.textSize
    merged.textColor = safeParseColor(customization.textColor) ?: this.textColor
    merged.fontResId = customization.font?.toFontResId(context)
    merged.contentSource = customization.contentSource?.mapToBlazeEnumClass() ?: this.contentSource
    merged.isVisible = customization.isVisible ?: this.isVisible

    return merged
}

fun BlazeMomentsPlayerBodyTextStyle.mergedWith(
    customization: BlazeReactMomentsPlayerBodyTextStyle?,
    context: Context
): BlazeMomentsPlayerBodyTextStyle {
    customization ?: return this

    val merged = this

    merged.textSize = customization.textSize ?: this.textSize
    merged.textColor = safeParseColor(customization.textColor) ?: this.textColor
    merged.fontResId = customization.font?.toFontResId(context)
    merged.contentSource = customization.contentSource?.mapToBlazeEnumClass() ?: this.contentSource
    merged.isVisible = customization.isVisible ?: this.isVisible

    return merged
}

fun BlazeMomentsPlayerButtonsStyle.mergedWith(
    customization: BlazeReactMomentsPlayerButtonsStyle?,
    context: Context
): BlazeMomentsPlayerButtonsStyle {
    customization ?: return this

    val merged = this
    merged.mute = merged.mute.mergeButtonThemes(customization.mute, context)
    merged.exit = merged.exit.mergeButtonThemes(customization.exit, context)
    merged.share = merged.share.mergeButtonThemes(customization.share, context)
    merged.like = merged.like.mergeButtonThemes(customization.like, context)
    merged.play = merged.play.mergeButtonThemes(customization.play, context)
    merged.seekForward = merged.seekForward.mergeButtonThemes(customization.seekForward, context)
    merged.seekBackward = merged.seekBackward.mergeButtonThemes(customization.seekBackward, context)
    merged.search = merged.search.mergeButtonThemes(customization.search, context)

    mergeCustomActionButtons(customization.customActionButtons, context)?.let {
        merged.setBottomStackCustomActionButtons(it)
    }

    return merged
}

private fun mergeCustomActionButtons(
    reactButtons: List<BlazeReactCustomActionButton>?,
    context: Context
): List<BlazeMomentsPlayerCustomActionButton>? {
    reactButtons ?: return null
    val nativeButtons = reactButtons.mapNotNull { reactButton ->
        val params = reactButton.customParams?.toNativeParams() ?: return@mapNotNull null
        val button = BlazeMomentsPlayerCustomActionButton(customParams = params)
        button.style = button.style.mergeButtonThemes(reactButton.style, context)
        button
    }
    return nativeButtons.ifEmpty { return null }
}

fun BlazeMomentsPlayerCtaStyle.mergedWith(
    customization: BlazeReactMomentsPlayerCtaStyle?,
    context: Context
): BlazeMomentsPlayerCtaStyle {
    customization ?: return this

    val merged = this
    merged.cornerRadius = customization.cornerRadius?.blazeDp ?: this.cornerRadius
    merged.textSize = customization.textSize ?: this.textSize
    merged.fontResId = customization.font?.toFontResId(context = context)
    merged.width = customization.width?.blazeDp ?: this.width
    merged.height = customization.height?.blazeDp ?: this.height
    customization.layoutPositioning?.let { merged.layoutPositioning = it.mapToBlazeEnumClass() }
    customization.horizontalAlignment?.let { merged.horizontalAlignment = it.mapToBlazeEnumClass() }
    customization.icon?.let { this.icon = this.icon.mergeWith(it, context) }
    merged.isVisible = customization.isVisible ?: this.isVisible
    return merged
}

fun BlazeMomentsPlayerCtaIconStyle?.mergeWith(
    customization: BlazeReactMomentsPlayerCtaIconStyle?,
    context: Context
): BlazeMomentsPlayerCtaIconStyle? {
    customization ?: return this
    val imageResId = customization.iconImage?.toImageResId(context)

    val iconPositioning = customization.iconPositioning?.mapToBlazeEnumClass()

    val merged = this

    merged?.let { iconStyle ->
        merged.iconTint = safeParseColor(customization.iconTint) ?: iconStyle.iconTint
        merged.iconPositioning = iconPositioning ?: iconStyle.iconPositioning
        merged.iconResId = imageResId ?: iconStyle.iconResId
    }
        ?: run {
            imageResId ?: return null
            iconPositioning ?: return null

            BlazeMomentsPlayerCtaIconStyle(
                iconResId = imageResId,
                iconPositioning = iconPositioning,
                iconTint = safeParseColor(customization.iconTint) ?: this?.iconTint
            )
        }

    return merged
}

fun BlazeMomentsPlayerChipsStyle.mergedWith(
    customization: BlazeReactMomentsPlayerChipsStyle?,
): BlazeMomentsPlayerChipsStyle {
    customization ?: return this

    val merged = this
    merged.ad = this.ad.mergedWith(customization.ad)
    return merged
}

fun BlazeMomentsPlayerFirstTimeSlideInstructionsStyle.mergedWith(
    customization: BlazeReactMomentsPlayerFirstTimeSlideInstructionsStyle?,
    context: Context
): BlazeMomentsPlayerFirstTimeSlideInstructionsStyle {
    customization ?: return this

    val merged = this
    merged.next = this.next.mergedWith(customization.next, context)
    merged.pause = this.pause.mergedWith(customization.pause, context)
    merged.play = this.play.mergedWith(customization.play, context)
    merged.previous = this.previous.mergedWith(customization.previous, context)
    return merged
}

fun BlazeMomentsPlayerSeekBarStyle.mergedWith(
    customization: BlazeReactMomentsPlayerSeekBarStyle?,
    context: Context
): BlazeMomentsPlayerSeekBarStyle {
    customization ?: return this

    val merged = this

    merged.isVisible = customization.isVisible ?: this.isVisible
    merged.playingState = merged.playingState.mergedWith(customization.playingState, context)
    merged.pausedState = merged.pausedState.mergedWith(customization.pausedState, context)
    merged.horizontalMargin = customization.horizontalSpacing?.blazeDp ?: merged.horizontalMargin
    merged.bottomMargin = customization.bottomSpacing?.blazeDp ?: this.bottomMargin

    return merged
}

fun BlazeMomentsPlayerChipStyle.mergedWith(
    customization: BlazeReactMomentsPlayerChipStyle?,
): BlazeMomentsPlayerChipStyle {
    customization ?: return this

    val merged = this
    merged.padding = this.padding.mergedWith(customization.titlePadding)
    merged.text = customization.text ?: this.text
    merged.textColor = safeParseColor(customization.textColor) ?: this.textColor
    merged.backgroundColor = safeParseColor(customization.backgroundColor) ?: this.backgroundColor
    merged.isVisible = customization.isVisible ?: this.isVisible

    return merged
}

fun BlazeMomentsPlayerHeaderGradientStyle.mergedWith(
    customization: BlazeReactMomentsPlayerHeaderGradientStyle?,
): BlazeMomentsPlayerHeaderGradientStyle {
    customization ?: return this

    val merged = this
    merged.isVisible = customization.isVisible ?: this.isVisible
    merged.startColor = safeParseColor(customization.startColor) ?: this.startColor
    merged.endColor = safeParseColor(customization.endColor) ?: this.endColor

    return merged
}

fun BlazeMomentsPlayerFooterGradientStyle.mergedWith(
    customization: BlazeReactMomentsPlayerFooterGradientStyle?,
): BlazeMomentsPlayerFooterGradientStyle {
    customization ?: return this

    val merged = this
    customization.isVisible?.let { merged.isVisible = it }
    safeParseColor(customization.startColor)?.let { merged.startColor = it }
    safeParseColor(customization.endColor)?.let { merged.endColor = it }
    customization.endPositioning?.mapToBlazeEnumClass()?.let { merged.endPositioning = it }
    return merged
}

File: blaze_flutter_sdk/ios/blaze_flutter_sdk/Sources/Players/Moments/BlazeReactMomentsPlayerStyleMerger.swift
//
//  BlazeReactMomentsPlayerStyleMerger.swift
//  blaze_flutter_sdk
//
//

import BlazeSDK
import Foundation
import UIKit

extension BlazeMomentsPlayerStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.headingText = merged.headingText.mergedWith(customization.headingText)
        merged.bodyText = merged.bodyText.mergedWith(customization.bodyText)
        merged.buttons = merged.buttons.mergedWith(customization.buttons)
        merged.chips = merged.chips.mergedWith(customization.chips)
        merged.cta = merged.cta.mergedWith(customization.cta)
        merged.headerGradient = merged.headerGradient.mergedWith(customization.headerGradient)
        merged.footerGradient = merged.footerGradient.mergedWith(customization.footerGradient)
        merged.backgroundColor =
            UIColor(hexString: customization.backgroundColor) ?? merged.backgroundColor
        merged.firstTimeSlide = merged.firstTimeSlide.mergedWith(customization.firstTimeSlide)
        merged.seekBar = merged.seekBar.mergedWith(customization.seekBar)
        merged.bottomComponentsAlignment = merged.bottomComponentsAlignment.mergedWith(
            customization.bottomComponentsAlignment)
        merged.playerDisplayMode = merged.playerDisplayMode.mergedWith(
            customization.playerDisplayMode)
        merged.followEntity = merged.followEntity.mergedWith(customization.followEntity)
        merged.captions = merged.captions.mergedWith(customization.captions)
        return merged
    }
}

extension BlazeMomentsPlayerFollowEntityStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerFollowEntityStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.followState = merged.followState.mergedWith(customization.followState)
        merged.unfollowState = merged.unfollowState.mergedWith(customization.unfollowState)
        merged.entityType = merged.entityType.mergedWith(customization.entityType)
        return merged
    }
}

extension BlazeMomentsPlayerFollowEntityStateStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerFollowEntityStateStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.avatar = merged.avatar.mergedWith(customization.avatar)
        merged.chip = merged.chip.mergedWith(customization.chip)
        return merged
    }
}

extension BlazeMomentsPlayerFollowEntityAvatarStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerFollowEntityAvatarStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.borderWidth = customization.borderWidth ?? merged.borderWidth
        merged.borderColor = UIColor(hexString: customization.borderColor) ?? merged.borderColor
        return merged
    }
}

extension BlazeMomentsPlayerFollowEntityChipStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerFollowEntityChipStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.backgroundColor =
            UIColor(hexString: customization.backgroundColor) ?? merged.backgroundColor
        merged.iconColor = UIColor(hexString: customization.iconColor) ?? merged.iconColor
        merged.contentSource = merged.contentSource.mergedWith(customization.contentSource)
        return merged
    }
}

extension BlazeMomentsPlayerFollowEntityChipContentSource {
    func mergedWith(_ customization: BlazeReactMomentsPlayerFollowEntityChipStyle.ContentSource?)
        -> Self
    {
        guard let customization else { return self }

        switch customization {
        case .icon:
            return .icon
        case .text:
            return .text
        }
    }
}

extension BlazeFollowEntityType {
    func mergedWith(_ customization: BlazeReactFollowEntityType?) -> Self {
        guard let customization, customization.type != nil else { return self }
        return Self.fromReact(customization) ?? self
    }

    private static func fromReact(_ react: BlazeReactFollowEntityType) -> Self? {
        guard let type = react.type else { return nil }
        let fallback = react.fallbackType.flatMap { fromReact($0) }

        switch type {
        case .firstAvailable:
            return .firstAvailable
        case .player:
            return .player(fallbackType: fallback)
        case .team:
            return .team(fallbackType: fallback)
        case .property:
            return .property(fallbackType: fallback)
        }
    }
}

extension BlazeMomentsPlayerSeekBarStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerSeekBarStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.playingState = merged.playingState.mergedWith(customization.playingState)
        merged.pausedState = merged.pausedState.mergedWith(customization.pausedState)
        merged.horizontalSpacing = customization.horizontalSpacing ?? merged.horizontalSpacing
        merged.bottomSpacing = customization.bottomSpacing ?? merged.bottomSpacing
        return merged
    }
}

extension BlazeMomentsPlayerFirstTimeSlideInstructionsStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerFirstTimeSlideInstructionsStyle?)
        -> Self
    {
        guard let customization else { return self }

        var merged = self
        merged.next = merged.next.mergedWith(customization.next)
        merged.pause = merged.pause.mergedWith(customization.pause)
        merged.play = merged.play.mergedWith(customization.play)
        merged.previous = merged.previous.mergedWith(customization.previous)
        return merged
    }
}

extension BlazeMomentsPlayerFirstTimeSlideStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerFirstTimeSlideStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.backgroundColor =
            UIColor(hexString: customization.backgroundColor?.colorName) ?? merged.backgroundColor
        merged.show = customization.show ?? merged.show
        merged.mainTitle = merged.mainTitle.mergedWith(customization.mainTitle)
        merged.cta = merged.cta.mergedWith(customization.cta)
        merged.subtitle = merged.subtitle.mergedWith(customization.subtitle)
        merged.instructions = merged.instructions.mergedWith(customization.instructions)
        return merged
    }
}

extension BlazeMomentsPlayerHeaderGradientStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerHeaderGradientStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.startColor = UIColor(hexString: customization.startColor) ?? merged.startColor
        merged.endColor = UIColor(hexString: customization.endColor) ?? merged.endColor
        return merged
    }
}

extension BlazeMomentsPlayerFooterGradientStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerFooterGradientStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.startColor = UIColor(hexString: customization.startColor) ?? merged.startColor
        merged.endColor = UIColor(hexString: customization.endColor) ?? merged.endColor
        merged.endPositioning = merged.endPositioning.mergedWith(customization.endPositioning)
        return merged
    }
}

extension BlazeMomentsPlayerFooterGradientStyle.EndPositioning {

    func mergedWith(_ customization: BlazeReactMomentsPlayerFooterGradientStyle.EndPositioning?)
        -> Self
    {
        guard let customization else { return self }

        switch customization {
        case .bottomToContainer:
            return .bottomToContainer
        case .bottomToPlayer:
            return .bottomToPlayer
        }
    }

}

extension BlazeMomentsPlayerCtaStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerCtaStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.cornerRadius = customization.cornerRadius ?? merged.cornerRadius
        merged.height = customization.height ?? merged.height
        merged.width = customization.width ?? merged.width
        merged.font = merged.font.fontWith(customization.font, textSize: customization.textSize)
        merged.layoutPositioning = merged.layoutPositioning.mergedWith(
            customization.layoutPositioning)
        merged.horizontalAlignment = merged.horizontalAlignment.mergedWith(
            customization.horizontalAlignment)
        merged.icon = UIImage.imageWith(customization.icon?.iconImage)
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.borderColor = UIColor(hexString: customization.borderColor) ?? merged.borderColor
        merged.borderWidth = customization.borderWidth ?? merged.borderWidth
        return merged
    }
}

extension BlazeMomentsPlayerCtaStyle.CTAPositioning {
    func mergedWith(_ customization: BlazeReactMomentsPlayerCtaStyle.CTAPositioning?) -> Self {
        guard let customization else { return self }

        switch customization {
        case .bellowBottomButtonsBox:
            return .ctaBellowBottomButtonsBox
        case .nextToBottomButtonsBox:
            return .ctaNextToBottomButtonsBox
        }

    }
}

extension BlazeMomentsPlayerCtaStyle.CTAHorizontalAlignment {
    func mergedWith(_ customization: BlazeReactMomentsPlayerCtaStyle.CTAHorizontalAlignment?)
        -> Self
    {
        guard let customization else { return self }

        switch customization {
        case .start:
            return .leading
        case .center:
            return .center
        case .end:
            return .trailing
        case .fullAvailableWidth:
            return .fullAvailableWidth
        }
    }
}

extension BlazeMomentsPlayerChipsStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerChipsStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.ad = merged.ad.mergedWith(customization.ad)
        return merged
    }
}

extension BlazeMomentsPlayerChipStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerChipStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.backgroundColor =
            UIColor(hexString: customization.backgroundColor) ?? merged.backgroundColor
        merged.text = customization.text ?? merged.text
        merged.textColor = UIColor(hexString: customization.textColor) ?? merged.textColor
        merged.titlePadding = merged.titlePadding.mergedWith(customization.titlePadding)
        merged.isVisible = customization.isVisible ?? merged.isVisible
        return merged
    }
}

extension BlazeMomentsPlayerButtonsStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerButtonsStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.exit = merged.exit.mergedWith(customization.exit)
        merged.like = merged.like.mergedWith(customization.like)
        merged.mute = merged.mute.mergedWith(customization.mute)
        merged.play = merged.play.mergedWith(customization.play)
        merged.share = merged.share.mergedWith(customization.share)
        merged.captions = merged.captions.mergedWith(customization.captions)
        merged.seekForward = merged.seekForward.mergedWith(customization.seekForward)
        merged.seekBackward = merged.seekBackward.mergedWith(customization.seekBackward)
        merged.search = merged.search.mergedWith(customization.search)

        if let customActionButtons = customization.customActionButtons, !customActionButtons.isEmpty {
            let nativeButtons = customActionButtons.compactMap { $0.toNativeMomentsCustomActionButton() }
            if !nativeButtons.isEmpty {
                merged.setBottomStackCustomActionButtons(nativeButtons)
            }
        }

        merged.shouldPreserveLikesCountBelowThreshold =
            customization.shouldPreserveLikesCountBelowThreshold
            ?? merged.shouldPreserveLikesCountBelowThreshold

        return merged
    }
}

extension BlazeMomentsPlayerButtonStyle {
    func mergedWith(_ customization: BlazeReactPlayerButtonStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.color = UIColor(hexString: customization.color) ?? merged.color
        merged.height = customization.height ?? merged.height
        merged.width = customization.width ?? merged.width
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.isVisibleForAds = customization.isVisibleForAds ?? merged.isVisibleForAds
        merged.customImage = merged.customImage.mergedWith(customization.customImage)
        merged.contentHorizontalAlignment = .center
        merged.contentVerticalAlignment = .center

        return merged
    }
}

extension BlazeMomentsPlayerHeadingTextStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerHeadingTextStyle?) -> Self {
        guard let customization else { return self }

        var merged = self

        merged.textColor = UIColor(hexString: customization.textColor) ?? merged.textColor
        merged.font = merged.font.fontWith(customization.font, textSize: customization.textSize)
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.contentSource = merged.contentSource.mergedWith(customization.contentSource)

        return merged
    }
}

extension BlazeMomentsPlayerBodyTextStyle {
    func mergedWith(_ customization: BlazeReactMomentsPlayerBodyTextStyle?) -> Self {
        guard let customization else { return self }

        var merged = self

        merged.textColor = UIColor(hexString: customization.textColor) ?? merged.textColor
        merged.font = merged.font.fontWith(customization.font, textSize: customization.textSize)
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.contentSource = merged.contentSource.mergedWith(customization.contentSource)

        return merged
    }
}

extension BlazeMomentsPlayerHeadingTextStyle.ContentSource {

    func mergedWith(_ customization: BlazeReactMomentsPlayerHeadingTextStyle.ContentSource?) -> Self
    {
        guard let customization else { return self }

        switch customization {
        case .title:
            return .title
        case .subtitle:
            return .subtitle
        }
    }

}

extension BlazeMomentsPlayerBodyTextStyle.ContentSource {

    func mergedWith(_ customization: BlazeReactMomentsPlayerBodyTextStyle.ContentSource?) -> Self {
        guard let customization else { return self }

        switch customization {
        case .title:
            return .title
        case .subtitle:
            return .subtitle
        case .description:
            return .description
        }
    }

}
extension BlazeMomentsPlayerStyle.BottomComponentsAlignment {

    func mergedWith(_ customization: BlazeReactBottomComponentsAlignment?) -> Self {
        guard let customization else { return self }

        switch customization {
        case .relativeToContainer:
            return .relativeToContainer
        case .relativeToPlayer:
            return .relativeToPlayer
        case .fitCtaBelowPlayer:
            return .fitCTABelowPlayer
        }
    }

}

extension BlazePlayerDisplayMode {

    func mergedWith(_ customization: BlazeReactPlayerDisplayMode?) -> Self {
        guard let customization else { return self }

        switch customization {
        case .fixedRatio9x16:
            return .fixedRatio_9_16
        case .resizeAspectFillCenterCrop:
            return .resizeAspectFillCenterCrop
        }
    }
}

extension BlazeSDK.BlazeMomentsPlayerButtonCustomImageStates? {

    func mergedWith(_ customization: BlazeReactPlayerButtonCustomImageStates?) -> Self {
        guard let customization else { return self }

        guard let defaultImage = UIImage.imageWith(customization.unselectedImage) else {
            return self
        }

        return .init(
            default: defaultImage,
            selected: UIImage.imageWith(customization.selectedImage))
    }

}
